In [6]:
%load_ext sql

In [7]:
%sql postgresql://appdev@data/appdev

'Connected: appdev@appdev'

In [3]:
# Printing table
%sql SELECT * FROM pg_catalog.pg_tables;

136 rows affected.


[('chinook', 'invoice', 'appdev', None, True, False, True, False),
 ('chinook', 'customer', 'appdev', None, True, False, True, False),
 ('chinook', 'mediatype', 'appdev', None, True, False, True, False),
 ('chinook', 'playlist', 'appdev', None, True, False, True, False),
 ('eav', 'support', 'appdev', None, True, False, True, False),
 ('chinook', 'genre', 'appdev', None, True, False, True, False),
 ('chinook', 'invoiceline', 'appdev', None, True, False, True, False),
 ('pg_catalog', 'pg_statistic', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_type', 'postgres', None, True, False, False, False),
 ('chinook', 'track', 'appdev', None, True, False, True, False),
 ('chinook', 'playlisttrack', 'appdev', None, True, False, True, False),
 ('eav', 'customer', 'appdev', None, True, False, False, False),
 ('eav', 'params', 'appdev', None, True, False, False, False),
 ('chinook', 'employee', 'appdev', None, True, False, True, False),
 ('chinook', 'artist', 'appdev', None, True, False, True, False),
 ('pg_catalog', 'pg_policy', 'postgres', None, True, False, False, False),
 ('f1db', 'circuits', 'appdev', None, True, False, False, False),
 ('eav', 'support_contract_type', 'appdev', None, True, False, True, False),
 ('pg_catalog', 'pg_authid', 'postgres', 'pg_global', True, False, False, False),
 ('f1db', 'constructorresults', 'appdev', None, True, False, False, False),
 ('f1db', 'constructors', 'appdev', None, True, False, False, False),
 ('f1db', 'constructorstandings', 'appdev', None, True, False, False, False),
 ('f1db', 'drivers', 'appdev', None, True, False, False, False),
 ('f1db', 'driverstandings', 'appdev', None, True, False, False, False),
 ('f1db', 'laptimes', 'appdev', None, True, False, False, False),
 ('f1db', 'pitstops', 'appdev', None, True, False, False, False),
 ('f1db', 'qualifying', 'appdev', None, True, False, False, False),
 ('f1db', 'races', 'appdev', None, True, False, False, False),
 ('f1db', 'results', 'appdev', None, True, False, False, False),
 ('f1db', 'seasons', 'appdev', None, True, False, False, False),
 ('f1db', 'status', 'appdev', None, True, False, False, False),
 ('geoname', 'class', 'appdev', None, True, False, True, False),
 ('geoname', 'feature', 'appdev', None, True, False, True, False),
 ('geoname', 'district', 'appdev', None, True, False, True, False),
 ('pg_catalog', 'pg_user_mapping', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_subscription', 'postgres', 'pg_global', True, False, False, False),
 ('geoname', 'continent', 'appdev', None, True, False, True, False),
 ('geoname', 'geoname', 'appdev', None, True, False, True, False),
 ('pg_catalog', 'pg_attribute', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_proc', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_class', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_attrdef', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_constraint', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_inherits', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_index', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_operator', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_opfamily', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_opclass', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_am', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_amop', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_amproc', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_language', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_largeobject_metadata', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_aggregate', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_statistic_ext', 'postgres', None, True, False, False, False),
 ('pg_catalog', 'pg_rewrite', 'postgres', None, 

#### 1 - On the table circuits report:
- What type of indices exists for the table and why they are of that type (and not some other type)
- The amount of space each index takes up

In [28]:
# What type of indices exists for the table and why they are of that type (and not some other type)
%sql SELECT * FROM pg_indexes WHERE tablename = 'circuits';

3 rows affected.


[('f1db', 'circuits', 'idx_17102_url', None, 'CREATE UNIQUE INDEX idx_17102_url ON circuits USING btree (url)'),
 ('f1db', 'circuits', 'circuits_position_idx', None, 'CREATE INDEX circuits_position_idx ON circuits USING gist ("position")'),
 ('f1db', 'circuits', 'idx_17102_primary', None, 'CREATE UNIQUE INDEX idx_17102_primary ON circuits USING btree (circuitid)')]

What type of indices exists for the table and why they are of that type (and not some other type):

1. btree (url)
    - Sorting by alphabetic order. Extension of the binary tree, where you can have more than two leaves.
2. gist ("position")
    - Generalizes search structures, so we can build them on whicheever type we choose. Contains a balanced tree structure. <key, pointer> pairs instead of integers (like in B-trees). 
3. btree (circuitid)
    - Sorting by numbers instead of by characters like the url btree.
    
The amount of space each index takes up:

In [36]:
# idx_17102_url
%sql select pg_size_pretty(pg_relation_size('idx_17102_url'));

1 rows affected.


[('16 kB',)]

In [37]:
# circuits_position_idx
%sql select pg_size_pretty(pg_relation_size('circuits_position_idx'));

1 rows affected.


[('8192 bytes',)]

In [38]:
# idx_17102_primary
%sql select pg_size_pretty(pg_relation_size('idx_17102_primary	'));

1 rows affected.


[('16 kB',)]

The three above queries could also be done in terminal by writing: 

- appdev-> \di+ "indexname"; 

Forexample:

#### 2 - We are talent scouts looking to win over some of the best new drivers there are. But we don't want them too old! Write a query that finds the winner of all the races, but only if they are younger than 38 years. The query should give return the date, driver surname, driver age, track time in milliseconds, race name and circuit name for all races.
In order to solve this assignment we will need to
 join multiple tables: 

Circuits
- (name) circuit name

Drivers
- (surname) driver surname
- (dob) drivers age

Races
- (date) race date
- (name) race name

results
- (milliseconds) track time in milliseconds

In [50]:
%sql SELECT races.date, drivers.surname, drivers.dob, results.milliseconds, races.name racename, circuits.name circuitname FROM results JOIN races using (raceid) JOIN drivers using (driverid) JOIN circuits using (circuitid) WHERE dob >= '1980-01-01' AND rank = 1;

162 rows affected.


[(datetime.date(2008, 3, 16), 'Kovalainen', datetime.date(1981, 10, 19), 5708630, 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2008, 4, 6), 'Kovalainen', datetime.date(1981, 10, 19), 5493759, 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2008, 8, 24), 'Massa', datetime.date(1981, 4, 25), 5732339, 'European Grand Prix', 'Valencia Street Circuit'),
 (datetime.date(2008, 10, 12), 'Massa', datetime.date(1981, 4, 25), 5468050, 'Japanese Grand Prix', 'Fuji Speedway'),
 (datetime.date(2008, 10, 19), 'Hamilton', datetime.date(1985, 1, 7), 5517403, 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2008, 11, 2), 'Massa', datetime.date(1981, 4, 25), 5651435, 'Brazilian Grand Prix', 'Autódromo José Carlos Pace'),
 (datetime.date(2007, 4, 8), 'Hamilton', datetime.date(1985, 1, 7), 5552487, 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2007, 4, 15), 'Massa', datetime.date(1981, 4, 25), 5607515, 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2007, 5, 13), 'Massa', datetime.date(1981, 4, 25), 5496230, 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya'),
 (datetime.date(2007, 5, 27), 'Alonso', datetime.date(1981, 7, 29), 6029329, 'Monaco Grand Prix', 'Circuit de Monaco'),
 (datetime.date(2007, 6, 10), 'Alonso', datetime.date(1981, 7, 29), 6273228, 'Canadian Grand Prix', 'Circuit Gilles Villeneuve'),
 (datetime.date(2007, 7, 1), 'Massa', datetime.date(1981, 4, 25), 5456614, 'French Grand Prix', 'Circuit de Nevers Magny-Cours'),
 (datetime.date(2007, 7, 22), 'Massa', datetime.date(1981, 4, 25), 7594513, 'European Grand Prix', 'Nürburgring'),
 (datetime.date(2007, 9, 9), 'Alonso', datetime.date(1981, 7, 29), 4717806, 'Italian Grand Prix', 'Autodromo Nazionale di Monza'),
 (datetime.date(2007, 9, 16), 'Massa', datetime.date(1981, 4, 25), 4843761, 'Belgian Grand Prix', 'Circuit de Spa-Francorchamps'),
 (datetime.date(2007, 9, 30), 'Hamilton', datetime.date(1985, 1, 7), 7234579, 'Japanese Grand Prix', 'Fuji Speedway'),
 (datetime.date(2007, 10, 7), 'Massa', datetime.date(1981, 4, 25), 5891286, 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2006, 3, 12), 'Rosberg', datetime.date(1985, 6, 27), 5449248, 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2006, 3, 19), 'Alonso', datetime.date(1981, 7, 29), 5445114, 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2006, 4, 23), 'Alonso', datetime.date(1981, 7, 29), 5468582, 'San Marino Grand Prix', 'Autodromo Enzo e Dino Ferrari'),
 (datetime.date(2006, 5, 14), 'Massa', datetime.date(1981, 4, 25), 5211612, 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya'),
 (datetime.date(2006, 6, 11), 'Alonso', datetime.date(1981, 7, 29), 5151927, 'British Grand Prix', 'Silverstone Circuit'),
 (datetime.date(2006, 8, 6), 'Massa', datetime.date(1981, 4, 25), None, 'Hungarian Grand Prix', 'Hungaroring'),
 (datetime.date(2006, 10, 1), 'Alonso', datetime.date(1981, 7, 29), 5855868, 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2006, 10, 8), 'Alonso', datetime.date(1981, 7, 29), 5032413, 'Japanese Grand Prix', 'Suzuka Circuit'),
 (datetime.date(2005, 3, 6), 'Alonso', datetime.date(1981, 7, 29), 5064048, 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2005, 5, 29), 'Alonso', datetime.date(1981, 7, 29), 5506648, 'European Grand Prix', 'Nürburgring'),
 (datetime.date(2009, 3, 29), 'Rosberg', datetime.date(1985, 6, 27), 5661506, 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2009, 4, 5), 'Button', datetime.date(1980, 1, 19), 4252092, 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2009, 5, 24), 'Massa', datetime.date(1981, 4, 25), 6059392, 'Monaco Grand Prix', 'Circuit de Monaco'),
 (datetime.date(2009, 6, 7), 'Button', datetime.date(1980, 1, 19), 5184848, 'Turkish Grand Prix', 'Istanbul Park'),
 (datetime.date(20

#### 3 - Describe the query using EXPLAIN ANALYZE with at least 5 lines of text. Answer at least the following:
- How many calls are you making?
- How long does it take to perform the query?


In [64]:
%sql EXPLAIN ANALYZE SELECT races.date, drivers.surname, drivers.dob, results.milliseconds, races.name racename, circuits.name circuitname FROM results JOIN races using (raceid) JOIN drivers using (driverid) JOIN circuits using (circuitid) WHERE dob >= '1980-01-01' AND rank = 1;

18 rows affected.


[('Nested Loop  (cost=25.68..752.35 rows=18 width=62) (actual time=0.242..4.785 rows=162 loops=1)',),
 ('  ->  Nested Loop  (cost=25.54..749.18 rows=18 width=50) (actual time=0.237..4.503 rows=162 loops=1)',),
 ('        ->  Hash Join  (cost=25.26..737.47 rows=18 width=27) (actual time=0.227..4.132 rows=162 loops=1)',),
 ('              Hash Cond: (results.driverid = drivers.driverid)',),
 ('              ->  Seq Scan on results  (cost=0.00..708.96 rows=257 width=24) (actual time=0.016..3.846 rows=257 loops=1)',),
 ('                    Filter: (rank = 1)',),
 ('                    Rows Removed by Filter: 23420',),
 ('              ->  Hash  (cost=24.51..24.51 rows=60 width=19) (actual time=0.202..0.202 rows=61 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                    ->  Seq Scan on drivers  (cost=0.00..24.51 rows=60 width=19) (actual time=0.008..0.185 rows=61 loops=1)',),
 ("                          Filter: (dob >= '1980-01-01'::date)",),
 ('                          Rows Removed by Filter: 780',),
 ('        ->  Index Scan using idx_17199_primary on races  (cost=0.28..0.65 rows=1 width=39) (actual time=0.002..0.002 rows=1 loops=162)',),
 ('              Index Cond: (raceid = results.raceid)',),
 ('  ->  Index Scan using idx_17102_primary on circuits  (cost=0.14..0.18 rows=1 width=28) (actual time=0.001..0.001 rows=1 loops=162)',),
 ('        Index Cond: (circuitid = races.circuitid)',),
 ('Planning time: 0.487 ms',),
 ('Execution time: 4.846 ms',)]

How many calls are you making? 
- 7 (not particulary sure about this answer, should I perhabs analyse the query plan above and multply the amount of rows that is being looped though, etc?)

How long does it take to perform the query?
- Planning time: 0.436 ms
- Execution time: 4.330 ms

First of all what does the EXPLAIN command mean? It is a way of showing the execution plan of a statement/query. The command displays the execution plan that the PostgreSQL planner generates for the supplied statement. The ANALYZE command makes executes the statement. The query takes this amount of time to run because it has to execute everything. It has to join all the different tables by specified using and then it outprints the specified tables. Nothing is stored in cache it has to be run everytime, which makes it really slow and CPU heavy. In order to speed up the process you would create a materialized view, that would be stored static as data in cache. 

#### 4 - Create a materialized view of your query. Using EXPLAIN ANALYZE try to query the view. Write at least 5 lines of text explaining what's going on and why the query execution time changed.

In [68]:
# Creating the view
# %sql CREATE MATERIALIZED VIEW best_racers_under_38 AS SELECT races.date, drivers.surname, drivers.dob, results.milliseconds, races.name racename, circuits.name circuitname FROM results JOIN races using (raceid) JOIN drivers using (driverid) JOIN circuits using (circuitid) WHERE dob >= '1980-01-01' AND rank = 1;

# EXPLAINING and ANALYZING the view
%sql EXPLAIN ANALYZE SELECT * FROM best_racers_under_38;

3 rows affected.


[('Seq Scan on best_racers_under_38  (cost=0.00..4.62 rows=162 width=66) (actual time=0.012..0.025 rows=162 loops=1)',),
 ('Planning time: 0.042 ms',),
 ('Execution time: 0.048 ms',)]

The queries execution time changed from 4.330 ms to 0.048, which is a drastic change! Why has this happend? This is because when using Materialised Views it is stored in cache. This is great if your data will not be changing anytime soon, because Materialised Views are static snapshots, therefore they will not update if the data is updated. This is great if you have static data, but otherwise not. You will have to refresh the materialised view onces again in order to update it. Because of the view being static this is why the execution time is this much lower. The view is executing faster because the data is static and is stored in the cache. 